File: dotnet/src/FleetLink.Api/Services/VehicleService.cs
using FleetLink.Api.Common;
using FleetLink.Api.Data;
using FleetLink.Api.Dtos;
using FleetLink.Api.Models;

namespace FleetLink.Api.Services;

public sealed class VehicleService : IVehicleService
{
    private readonly FleetStore _store;
    public VehicleService(FleetStore store) => _store = store;

    public IReadOnlyList<VehicleDto> ListVehicles() =>
        _store.Vehicles.Select(v => v.ToDto()).ToList();

    public VehicleDto? GetVehicle(Guid id) =>
        _store.Vehicles.FirstOrDefault(v => v.Id == id)?.ToDto();

    public IReadOnlyList<WorkOrderDto> ListWorkOrdersForVehicle(Guid vehicleId)
    {
        if (_store.Vehicles.All(v => v.Id != vehicleId)) throw Errors.VehicleNotFound();      // rule 2
        return _store.WorkOrders
            .Where(w => w.VehicleId == vehicleId)
            .Select(w => w.ToDto(Costing.PartsCostFor(_store, w.Id)))
            .ToList();
    }

    // FSD rule 12 — odometer may only stay the same or increase.
    public VehicleDto UpdateOdometer(Guid vehicleId, int odometerKm)
    {
        var v = _store.Vehicles.FirstOrDefault(x => x.Id == vehicleId) ?? throw Errors.VehicleNotFound(); // rule 2
        if (odometerKm < v.OdometerKm) throw Errors.OdometerDecrease();                        // rule 12
        v.OdometerKm = odometerKm;
        return v.ToDto();
    }

    // FSD rule 7 — derive the vehicle's status from its open breakdown work, without ever touching a
    // Retired vehicle. Called after any create/status change that could open or close a Breakdown.
    public void RecomputeBreakdownStatus(Guid vehicleId)
    {
        var v = _store.Vehicles.FirstOrDefault(x => x.Id == vehicleId);
        if (v is null || v.Status == VehicleStatus.Retired) return;
        var hasOpenBreakdown = _store.WorkOrders.Any(w =>
            w.VehicleId == vehicleId && w.Type == WorkOrderType.Breakdown &&
            (w.Status == WorkOrderStatus.Open || w.Status == WorkOrderStatus.InProgress));
        v.Status = hasOpenBreakdown ? VehicleStatus.InMaintenance : VehicleStatus.Active;
    }
}

File: javascript/src/services/vehicleService.js
// Vehicle service (Module 2.D). Reads on the vehicle, plus the two vehicle-owned business rules:
//   • rule 12 — odometer is monotonic (may stay the same or increase, never decrease).
//   • rule 7  — breakdown auto-status: a vehicle is InMaintenance exactly while it has an Open/InProgress
//               Breakdown work order, and returns to Active when none remain. A Retired vehicle is never
//               auto-changed. This recompute is called by workOrderService whenever a breakdown opens or
//               closes, so the vehicle's status is always a consequence of its work, not set by hand.
import store from '../data/store.js';
import { Errors } from '../errors.js';
import { toVehicleDto, toWorkOrderDto } from '../dtos/mappers.js';
import { partsCostFor } from './costing.js';

const findVehicle = (id) => store.vehicles.find((v) => v.id === id);

export function listVehicles() {
  return store.vehicles.map(toVehicleDto);
}

export function getVehicle(id) {
  const v = findVehicle(id);
  return v ? toVehicleDto(v) : null;
}

export function listWorkOrdersForVehicle(vehicleId) {
  if (!findVehicle(vehicleId)) throw Errors.vehicleNotFound();               // rule 2 (distinguish 404 from empty)
  return store.workOrders
    .filter((w) => w.vehicleId === vehicleId)
    .map((w) => toWorkOrderDto(w, partsCostFor(w.id)));
}

// FSD rule 12 — odometer may only stay the same or increase.
export function updateOdometer(vehicleId, odometerKm) {
  const v = findVehicle(vehicleId);
  if (!v) throw Errors.vehicleNotFound();                                    // rule 2
  if (typeof odometerKm !== 'number' || odometerKm < v.odometerKm) throw Errors.odometerDecrease(); // rule 12
  v.odometerKm = odometerKm;
  return toVehicleDto(v);
}

// FSD rule 7 — derive the vehicle's status from its open breakdown work, without ever touching a
// Retired vehicle. Called after any create/status change that could open or close a Breakdown.
export function recomputeBreakdownStatus(vehicleId) {
  const v = findVehicle(vehicleId);
  if (!v || v.status === 'Retired') return;
  const hasOpenBreakdown = store.workOrders.some(
    (w) => w.vehicleId === vehicleId && w.type === 'Breakdown'
      && (w.status === 'Open' || w.status === 'InProgress'));
  v.status = hasOpenBreakdown ? 'InMaintenance' : 'Active';
}
